### ISSUE TITLE
YTM - Track artist matches previous track, not current

### ISSUE THREAD

**Issue by disjunto** [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-#3180]
**OS:** Windows

**GPMDP Version:** 4.6.0 with Youtube Music

**Issue Descriptions:** Current track is shown with artist from previous played track

**Steps to Reproduce:**

1. Start playing a playlist with multiple artists
2. Skip to next song
3. Check the song artist in either window title, or windows media control dialog
Expected: Artist matches the song
Actual: Artist matches the previous song in the playlist

**Comment by sharedito** [comment]
Also the track progression is not working, maybe same reason.

![image](https://user-images.githubusercontent.com/22381250/41870972-2ebde516-7883-11e8-938f-45fd3fa46178.png)

**Comment by dumontv** [comment]
Same issue on MacOS

**Comment by BastienMottier** [comment]
Same issue on Ubuntu

**Comment by HolyStephano** [comment]
[Having the same issues on Archlinux (Antergos)
](https://images.cloudninja.pw/PoliticalHurriedHeron.png)

**Comment by deanveloper** [comment]
Having same issue. Honestly quite annoying when I want to know who sings a song

**Comment by ctaetcsh** [comment]
This issue also affects Discord Rich Presence integration.

**Comment by mcgillowen** [comment]
Same issue on Ubuntu 18.04.

I believe I figured out why though. When there's a track change the YTM interface changes the track name nearly instantly but the the other information (Artist, Album, Like) is changed approximately half a second later. So the problem is that GPMDP scrapes the YTM interface instantly after the track changes but before everything excepting the track name has changed.

This can be checked by clicking the next track button while paying attention to the Artist name, it changes slightly after the track name. I've tested it in GPMDP as well as in Google Chrome.

I haven't had time to check where a change could be made, but a quick and ugly fix would be to update the artist and album information when the time gets updated. The best way to fix it would possibly be by scraping with a small delay after the track changes.

Hope this helps anybody looking into solving this problem, I'll try this weekend if I get the time.

**Comment by deanveloper** [comment]
> I haven't had time to check where a change could be made, but a quick and ugly fix would be to update the artist and album information when the time gets updated. The best way to fix it would possibly be by scraping with a small delay after the track changes.

I think "when the time gets updated" is a better solution. "with a small delay" implies that we know the delay is consistent, and that it won't change in the future.

**Comment by simoniz0r** [comment]
It should be relatively easy to code around this with some checks to see if the track name has changed.  Something like if new artist/time elapsed reset, but track not changed -> wait X seconds -> repeat check.

**Comment by IotaSpencer** [comment]
Alright considering the closing of that issue, whats the status of fixing this bug

**Comment by deanveloper** [comment]
I haven't heard of anyone making progress. We know the cause of it but idk if anyone is working on it.

**Comment by deanveloper** [comment]
I'm on my phone right now, once I'm home I'll fork and see if I can fix it

**Comment by jostrander** [comment]
If anyone hasn't tackled this yet, instead of using the html selectors in the YTM api, it may be possible to dive into the `__data__` on the `ytmusic-player-bar` node in the DOM. Cursory glancing shows all the data is present. 

```
// Album:
document.getElementsByTagName('ytmusic-player-bar')[0].__data__.playerPageWatchMetadata_.albumName.runs[0].text

// Track Title
document.getElementsByTagName('ytmusic-player-bar')[0].__data__.playerPageWatchMetadata_.title.runs[0].text

// Artist
document.getElementsByTagName('ytmusic-player-bar')[0].__data__.playerPageWatchMetadata_.byline.runs[0].text

// Year
document.getElementsByTagName('ytmusic-player-bar')[0].__data__.playerPageWatchMetadata_.byline.runs[4].text
```

**Comment by IotaSpencer** [comment]
` document.getElementsByTagName('ytmusic-player-bar')[0].__data__.playing_` for playing status, for state checking
also heres the object itself, from firefox
```json
{
  "title": {
    "runs": [
      {
        "text": "One More Time"
      }
    ]
  },
  "byline": {
    "runs": [
      {
        "text": "Daft Punk",
        "navigationEndpoint": {
          "clickTrackingParams": "CBMQkGsYACITCMPwldzxt98CFRnYnAod5VoJBg==",
          "browseEndpoint": {
            "browseId": "UCRr1xG_2WIDs18a6cIiCxeA"
          },
          "trackingParams": null
        }
      },
      {
        "text": " • "
      },
      {
        "text": "Discovery",
        "navigationEndpoint": {
          "clickTrackingParams": "CBMQkGsYACITCMPwldzxt98CFRnYnAod5VoJBg==",
          "musicEntityBrowseEndpoint": {
            "pageId": "DETAIL",
            "entityId": {
              "musicAlbumReleaseEntity": "Eg1iXzdsdE0zNGtyMG1IIBEoAQ%3D%3D"
            }
          },
          "trackingParams": null
        }
      },
      {
        "text": " • "
      },
      {
        "text": "2001"
      }
    ]
  },
  "secondaryByline": {
    "runs": [
      {
        "text": "2.4M views"
      },
      {
        "text": " • "
      },
      {
        "text": "41K likes"
      }
    ]
  },
  "secondaryTitle": {
    "runs": [
      {
        "text": "One More Time"
      }
    ]
  },
  "viewCountText": {
    "runs": [
      {
        "text": "2,499,015 views"
      }
    ]
  },
  "likeButton": {
    "likeButtonRenderer": {
      "target": {
        "videoId": "fa5IWHDbftI"
      },
      "likeStatus": "INDIFFERENT",
      "likeCount": 41096,
      "likeCountText": {
        "runs": [
          {
            "text": "41K"
          }
        ]
      },
      "likeCountWithLikeText": {
        "runs": [
          {
            "text": "41K"
          }
        ]
      },
      "dislikeCount": 8349,
      "dislikeCountText": {
        "runs": [
          {
            "text": "8.3K"
          }
        ]
      },
      "dislikeCountWithDislikeText": {
        "runs": [
          {
            "text": "8.3K"
          }
        ]
      },
      "trackingParams": "CBUQpUEiEwjD8JXc8bffAhUZ2JwKHeVaCQY=",
      "likesAllowed": true,
      "serviceEndpoints": [
        {
          "clickTrackingParams": "CBUQpUEiEwjD8JXc8bffAhUZ2JwKHeVaCQY=",
          "likeEndpoint": {
            "status": "LIKE",
            "target": {
              "videoId": "fa5IWHDbftI"
            }
          }
        },
        {
          "clickTrackingParams": "CBUQpUEiEwjD8JXc8bffAhUZ2JwKHeVaCQY=",
          "likeEndpoint": {
            "status": "DISLIKE",
            "target": {
              "videoId": "fa5IWHDbftI"
            }
          }
        },
        {
          "clickTrackingParams": "CBUQpUEiEwjD8JXc8bffAhUZ2JwKHeVaCQY=",
          "likeEndpoint": {
            "status": "INDIFFERENT",
            "target": {
              "videoId": "fa5IWHDbftI"
            }
          }
        }
      ],
      "dislikeNavigationEndpoint": {
        "clickTrackingParams": "CBUQpUEiEwjD8JXc8bffAhUZ2JwKHeVaCQYyDGRpc2xpa2VfbmV4dEjS_e2Gh4uS130=",
        "watchEndpoint": {
          "videoId": "yQBGdXGCUbA",
          "playlistId": "RDMMT3MP_RnlgwE",
          "index": 3,
          "params": "EAIYATgB"
        }
      }
    }
  },
  "trackingParams": "CBMQkGsYACITCMPwldzxt98CFRnYnAod5VoJBg==",
  "links": [
    {
      "compactLinkRenderer": {
        "icon": {
          "iconType": "ADD_TO_PLAYLIST"
        },
        "title": {
          "runs": [
            {
              "text": "Add to playlist"
            }
          ]
        },
        "trackingParams": "CBQQp0wYACITCMPwldzxt98CFRnYnAod5VoJBg==",
        "navigationEndpoint": {
          "clickTrackingParams": "CBQQp0wYACITCMPwldzxt98CFRnYnAod5VoJBg==",
          "addToPlaylistEndpoint": {
            "videoId": "fa5IWHDbftI"
          }
        }
      }
    }
  ],
  "albumName": {
    "runs": [
      {
        "text": "Discovery"
      }
    ]
  }
}
```

**Comment by IotaSpencer** [comment]
also @jostrander we need a position grabber, because the position that mpris reports is dead wrong

**Comment by IotaSpencer** [comment]
```
mpris:length
64-bit integer: The duration of the track in microseconds.
```
just looked at the specification

**Comment by IotaSpencer** [comment]
actually i can parse position and length fine. i just divide by 1,000,000 to get the amount of seconds

**Comment by jonyroda97** [comment]
Same here. Just occurs sometimes.

> **App Version:** 4.6.1
> **OS:** Windows 10 Pro x64 (1809)

**Comment by sargenthp** [comment]
Linux Mint 19 is also having the same issue.
About Google Play Music Desktop Player - Version: 4.6.1

![screenshot from 2019-01-25 08-43-35](https://user-images.githubusercontent.com/34111309/51752790-eebee700-207d-11e9-95a5-f5322f36865d.png)

Thanks!

**Comment by jostrander** [comment]
Fixed by #3586
